nsqd/topic.go

```diff
@@ -222,18 +220,34 @@ func (t *Topic) PutMessages(msgs []*Message) error {
 }
 
 func (t *Topic) put(m *Message) error {
-	select {
-	case t.memoryMsgChan <- m:
-	default:
-		err := writeMessageToBackend(m, t.backend)
-		t.nsqd.SetHealth(err)
-		if err != nil {
-			t.nsqd.logf(LOG_ERROR,
-				"TOPIC(%s) ERROR: failed to write message to backend - %s",
-				t.name, err)
-			return err
+	// if mem-queue-size == 0, avoid memory chan, for more consistent ordering
+	if cap(t.memoryMsgChan) > 0 {
+		select {
+		case t.memoryMsgChan <- m:
+			return nil
+		default:
+			break // write to backend
+		}
+	} else if t.ephemeral || m.deferred != 0 {
+		// try extra hard to deliver to channels through memory chan because
+		// messages lose deferred time in backend, and ephemeral topic has no backend
+		if len(t.channelMap) > 0 && !t.IsPaused() && t.IsStarted() {
+			select {
+			case t.memoryMsgChan <- m:
+				return nil
+			case <-time.After(time.Second):
+				break // give up
+			}
 		}
 	}
```

Comment:
I don't think we want to wait around for a second in the ephemeral case - the intention is that it's bounded and lossy, why add back pressure when the queue is full?

My proposal would be to simplify this whole if/else block to:

```go
if cap(t.memoryMsgChan) > 0 || m.deferred != 0 {
		select {
		case t.memoryMsgChan <- m:
			return nil
		default:
			break // write to backend
		}
}
```

Reply:
Yeah, we can pare it all the way down and remove the wait. The 1 second is arbitrary ... the expectation/desire is more like a handful of milliseconds. If there's a burst of messages and a bunch of channels, they can all get through the topic to the channels in just a few milliseconds, instead of having deferred time lost or messages kinda sampled.

Reply:
(oh and ephemeral topics don't work at all without the memory chan)